File: code/jvm/im-services/src/main/kotlin/im/services/channels/ChannelService.kt
package im.services.channels

import im.domain.Either
import im.domain.channel.Channel
import im.domain.channel.ChannelRole
import im.domain.user.User
import im.domain.wrappers.identifier.Identifier
import im.domain.wrappers.name.Name
import im.repository.pagination.Pagination
import im.repository.pagination.PaginationRequest
import im.repository.pagination.SortRequest
import im.services.users.UserError

interface ChannelService {
    /**
     * Creates a channel.
     *
     * - The channel name must be unique.
     * - The default role of the channel must be either [ChannelRole.MEMBER] or [ChannelRole.GUEST].
     *
     * @param name the name of the channel
     * @param defaultRole the default role of the channel for new members
     * @param isPublic whether the channel is public or private
     * @param user the user that is creating the channel
     * @return the created channel
     */
    fun createChannel(
        name: Name,
        defaultRole: ChannelRole,
        isPublic: Boolean,
        user: User,
    ): Either<ChannelError, Channel>

    /**
     * Retrieves a channel by its identifier.
     *
     * - A user must be a member of the channel to retrieve it.
     *
     * @param channelId the unique identifier of the channel
     * @param user the user that is requesting the channel
     * @return the channel if it exists, or a [ChannelError] otherwise
     */
    fun getChannelById(
        channelId: Identifier,
        user: User,
    ): Either<ChannelError, Channel>

    /**
     * Updates a channel.
     *
     * - A user must be the owner of the channel to update it.
     *
     * @param channelId the unique identifier of the channel
     * @param name the new name of the channel
     * @param defaultRole the new default role of the channel
     * @param isPublic whether the channel is public or private
     * @param user the user that is updating the channel
     * @return the updated channel
     */
    fun updateChannel(
        channelId: Identifier,
        name: Name,
        defaultRole: ChannelRole,
        isPublic: Boolean,
        user: User,
    ): Either<ChannelError, Unit>

    /**
     * Deletes a channel.
     *
     * - A user must be the owner of the channel to delete it.
     *
     * @param channelId the unique identifier of the channel
     * @param user the user that is deleting the channel
     */
    fun deleteChannel(
        channelId: Identifier,
        user: User,
    ): Either<ChannelError, Unit>

    /**
     * Joins a channel.
     *
     * - The channel must be public to join it directly.
     *
     * @param channelId the unique identifier of the channel
     * @param userId the unique identifier of the user
     * @param user the user that is joining the channel
     */
    fun joinChannel(
        channelId: Identifier,
        userId: Identifier,
        user: User,
    ): Either<ChannelError, Unit>

    /**
     * Removes a member from a channel.
     *
     * - A user must be the owner of the channel or be removing themselves to have permission to remove a member.
     *
     * @param channelId the unique identifier of the channel
     * @param userId the unique identifier of the user
     * @param user the user that is removing the member
     */
    fun removeChannelMember(
        channelId: Identifier,
        userId: Identifier,
        user: User,
    ): Either<ChannelError, Unit>

    /**
     * Retrieves a list of channels.
     *
     * @param name the name to search for
     * @param pagination the pagination request
     * @param sortRequest the sort request
     * @return a [Pagination] of channels if the search is successful, or a [ChannelError] otherwise
     */
    fun getChannels(
        name: String?,
        pagination: PaginationRequest,
        sortRequest: SortRequest,
    ): Either<ChannelError, Pagination<Channel>>

    /**
     * Retrieves the channels a user is a member of.
     *
     * @param userId the user identifier
     * @param user the authenticated user that is requesting the channels
     * @param sortRequest the sort request
     * @return a map of channels and their roles if the search is successful, or an [UserError] otherwise
     */
    fun getUserChannels(
        userId: Identifier,
        sortRequest: SortRequest,
        user: User,
    ): Either<ChannelError, Map<Channel, ChannelRole>>

    /**
     * Updates the role of a member in a channel.
     *
     * - An owner cannot change their own role.
     * - A user must be the owner of the channel to update the role of a member.
     * - The role of a member cannot be updated to owner.
     *
     * @param channelId the unique identifier of the channel
     * @param userId the unique identifier of the user
     * @param role the new role of the user
     *
     * @return a [ChannelError] if the update is unsuccessful, or [Unit] otherwise
     */
    fun updateMemberRole(
        channelId: Identifier,
        userId: Identifier,
        role: ChannelRole,
        user: User,
    ): Either<ChannelError, Unit>
}

File: code/js/src/Services/channels/ChannelService.ts
import { Name } from '../../Domain/wrappers/name/Name';
import { ChannelRole } from '../../Domain/channel/ChannelRole';
import { Session } from '../../Domain/sessions/Session';
import { ApiResult, handle } from '../media/Problem';
import {
    Channel,
    channelFromCreation,
    channelFromDto,
} from '../../Domain/channel/Channel';
import { BaseHTTPService } from '../BaseHTTPService';
import { ChannelCreationInputModel } from '../../Dto/input/ChannelCreationInputModel';
import { ChannelCreationOutputModel } from '../../Dto/output/channels/ChannelCreationOutputModel';
import { Identifier } from '../../Domain/wrappers/identifier/Identifier';
import { ChannelOutputModel } from '../../Dto/output/channels/ChannelOutputModel';
import { PaginationRequest } from '../../Domain/pagination/PaginationRequest';
import { SortRequest } from '../../Domain/pagination/SortRequest';
import { Pagination } from '../../Domain/pagination/Pagination';
import { ChannelsPaginatedOutputModel } from '../../Dto/output/channels/ChannelsPaginatedOutputModel';
import { ChannelRoleUpdateInputModel } from '../../Dto/input/ChannelRoleUpdateInputModel';
import { User } from '../../Domain/user/User';
import { Uri } from '../Uri';
import { buildQuery } from '../Utils';

export namespace ChannelService {
    import post = BaseHTTPService.post;
    import get = BaseHTTPService.get;
    import deleteRequest = BaseHTTPService.deleteRequest;
    import patch = BaseHTTPService.patch;

    import CHANNELS_ROUTE = Uri.CHANNELS_ROUTE;
    import CHANNEL_ROUTE = Uri.CHANNEL_ROUTE;
    import CHANNEL_ID_PARAM = Uri.CHANNEL_ID_PARAM;
    import CHANNEL_MEMBERS_ROUTE = Uri.CHANNEL_MEMBERS_ROUTE;
    import CHANNEL_MEMBER_ROUTE = Uri.CHANNEL_MEMBER_ROUTE;
    import USER_ID_PARAM = Uri.USER_ID_PARAM;
    import put = BaseHTTPService.put;

    /**
     * Creates a channel.
     *
     * @param name the name of the channel
     * @param defaultRole the default role of a user in the channel
     * @param isPublic indicates whether the channel is public
     * @param session the session of the user creating the channel
     * @param abortSignal the signal to abort the request
     *
     * @returns the channel that was created
     */
    export async function createChannel(
        name: Name,
        defaultRole: ChannelRole,
        isPublic: boolean,
        session: Session,
        abortSignal?: AbortSignal,
    ): ApiResult<Channel> {
        return handle(
            post<ChannelCreationInputModel, ChannelCreationOutputModel>({
                uri: CHANNELS_ROUTE,
                requestBody: {
                    name: name.value,
                    defaultRole: defaultRole,
                    isPublic: isPublic,
                },
                abortSignal: abortSignal,
            }),
            (outputModel) =>
                channelFromCreation(
                    outputModel,
                    name,
                    defaultRole,
                    session.user,
                    isPublic,
                ),
        );
    }

    /**
     * Gets a channel by its identifier.
     *
     * @param channelId the identifier of the channel
     * @param abortSignal the signal to abort the request
     *
     * @returns the channel
     */
    export async function getChannel(
        channelId: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<Channel> {
        return handle(
            get<ChannelOutputModel>({
                uri: CHANNEL_ROUTE.replace(
                    CHANNEL_ID_PARAM,
                    channelId.value.toString(),
                ),
                abortSignal: abortSignal,
            }),
            channelFromDto,
        );
    }

    /**
     * Gets a list of channels.
     *
     * @param name the name of the channel
     * @param pagination the pagination request
     * @param sort the sort request
     * @param abortSignal the signal to abort the request
     *
     * @returns the Channels with pagination information
     */
    export async function getChannels(
        name?: string,
        pagination?: PaginationRequest,
        sort?: SortRequest,
        abortSignal?: AbortSignal,
    ): ApiResult<Pagination<Channel>> {
        return handle(
            get<ChannelsPaginatedOutputModel>({
                uri: buildQuery(CHANNELS_ROUTE, name, pagination, sort),
                abortSignal: abortSignal,
            }),
            (outputModel) => ({
                items: outputModel.channels.map(channelFromDto),
                info: outputModel.pagination,
            }),
        );
    }

    /**
     * Updates a channel.
     *
     * @param channelId - the identifier of the channel
     * @param name - the name of the channel
     * @param defaultRole - the default role of a user in the channel
     * @param isPublic - indicates whether the channel is public
     * @param abortSignal - the signal to abort the request
     *
     * @returns the channel that was updated
     */
    export async function updateChannel(
        channelId: Identifier,
        name: Name,
        defaultRole: ChannelRole,
        isPublic: boolean,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return await put<ChannelCreationInputModel>({
            uri: CHANNEL_ROUTE.replace(
                CHANNEL_ID_PARAM,
                channelId.value.toString(),
            ),
            requestBody: {
                name: name.value,
                defaultRole: defaultRole,
                isPublic: isPublic,
            },
            fetchResBody: false,
            abortSignal: abortSignal,
        });
    }

    /**
     * Deletes a channel.
     *
     * @param channelId the identifier of the channel
     * @param abortSignal the signal to abort the request
     *
     * @returns the result of the request
     */
    export async function deleteChannel(
        channelId: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return await deleteRequest({
            uri: CHANNEL_ROUTE.replace(
                CHANNEL_ID_PARAM,
                channelId.value.toString(),
            ),
            abortSignal: abortSignal,
        });
    }

    /**
     * Adds the current session user to a channel.
     *
     * @param channel - the channel
     * @param session - the session
     * @param abortSignal - the signal to abort the request
     *
     * @returns the result of the request
     */
    export async function joinChannel(
        channel: Channel,
        session: Session,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return await put({
            uri: CHANNEL_MEMBERS_ROUTE.replace(
                CHANNEL_ID_PARAM,
                channel.id.value.toString(),
            ).replace(USER_ID_PARAM, session.user.id.value.toString()),
            abortSignal: abortSignal,
        });
    }

    /**
     * Removes a user from a channel.
     *
     * @param channel - the channel
     * @param userId - the identifier of the user
     * @param abortSignal - the signal to abort the request
     *
     * @returns the result of the request
     */
    export async function removeUserFromChannel(
        channel: Channel,
        userId: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return await deleteRequest({
            uri: CHANNEL_MEMBER_ROUTE.replace(
                CHANNEL_ID_PARAM,
                channel.id.value.toString(),
            ).replace(USER_ID_PARAM, userId.value.toString()),
            abortSignal: abortSignal,
        });
    }

    /**
     * Updates the role of a member in a channel.
     *
     * @param channel - the channel
     * @param user - the user
     * @param role - the role
     * @param abortSignal - the signal to abort the request
     *
     * @returns the result of the request
     */
    export async function updateMemberRole(
        channel: Channel,
        user: User,
        role: ChannelRole,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return await patch<ChannelRoleUpdateInputModel>({
            uri: CHANNEL_MEMBER_ROUTE.replace(
                CHANNEL_ID_PARAM,
                channel.id.value.toString(),
            ).replace(USER_ID_PARAM, user.id.value.toString()),
            requestBody: {
                role: role,
            },
            abortSignal: abortSignal,
        });
    }
}
